### PR TITLE
gh-103092: Isolate `_pickle`

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-103092

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2020-07-04-09-04-41.bpo-1635741.ZsP31Y.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Adapt :mod:`_pickle` to :pep:`687`. Patch by Mohamed Koubaa and Erlend Aasland.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2020-07-04-09-04-41.bpo-1635741.ZsP31Y.rst]:
```suggestion
Adapt :mod:`!_pickle` to :pep:`687`. Patch by Mohamed Koubaa and Erlend Aasland.
```

File: Modules/_pickle.c [hunk 2]
```diff
@@ -4546,21 +4578,23 @@ _pickle_Pickler_clear_memo_impl(PicklerObject *self)
...
 [clinic start generated code]*/
 
 static PyObject *
-_pickle_Pickler_dump(PicklerObject *self, PyObject *obj)
-/*[clinic end generated code: output=87ecad1261e02ac7 input=552eb1c0f52260d9]*/
+_pickle_Pickler_dump_impl(PicklerObject *self, PyTypeObject *cls,
+                          PyObject *obj)
+/*[clinic end generated code: output=952cf7f68b1445bb input=f949d84151983594]*/
 {
+    PickleState *st = _Pickle_GetStateByClass(cls);
```
Comment (kumaraditya303) [inline Modules/_pickle.c:4593]: `Pickler` has `Py_TPFLAGS_BASETYPE` set, is it safe to use `_Pickle_GetStateByClass`?

Comment (erlend-aasland) [reply Modules/_pickle.c:4593]: ~Good catch; no that definitely sounds unsafe.~

Looking twice; we use the defining class, so this is safe.

**Comment by erlend-aasland** [comment]
FYI, I made a PoC alternative variant where I store the state in the pickler and unpickler objects. It makes for a much nicer diff, but I had to do too much fishy stuff to make it work. You can take a look here: https://github.com/erlend-aasland/cpython/pull/new/isolate-pickle-alt

File: Modules/_pickle.c (hunk 2 above)
Comment (kumaraditya303) [reply Modules/_pickle.c:4593]: Right, I missed that.

File: Modules/_pickle.c [hunk 3]
```diff
@@ -4890,43 +4920,28 @@ PicklerMemoProxy_clear(PicklerMemoProxyObject *self)
...
-    0,                                          /* tp_richcompare */
-    0,                                          /* tp_weaklistoffset */
-    0,                                          /* tp_iter */
-    0,                                          /* tp_iternext */
-    picklerproxy_methods,                       /* tp_methods */
+static PyType_Slot memoproxy_slots[] = {
+    {Py_tp_dealloc, PicklerMemoProxy_dealloc},
+    {Py_tp_traverse, PicklerMemoProxy_traverse},
+    {Py_tp_clear, PicklerMemoProxy_clear},
+    {Py_tp_methods, picklerproxy_methods},
```
Comment (kumaraditya303) [inline Modules/_pickle.c:4927]: Missing `tp_hash = PyObject_HashNotImplemented`

File: Modules/_pickle.c [hunk 4]
```diff
@@ -4890,43 +4920,28 @@ PicklerMemoProxy_clear(PicklerMemoProxyObject *self)
...
+    .slots = memoproxy_slots,
 };
 
 static PyObject *
 PicklerMemoProxy_New(PicklerObject *pickler)
 {
     PicklerMemoProxyObject *self;
-
-    self = PyObject_GC_New(PicklerMemoProxyObject, &PicklerMemoProxyType);
+    PickleState *st = _Pickle_GetStateByClass(Py_TYPE(pickler));
```
Comment (kumaraditya303) [inline Modules/_pickle.c]: If `pickler` is a subclass then this will fail, you need to find state by module def here.

Comment (erlend-aasland) [reply Modules/_pickle.c]: True, good catch.

File: Modules/_pickle.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_pickle.c:4927]: Yep, true for both types.

File: Modules/_pickle.c (hunk 4 above)
Comment (erlend-aasland) [reply Modules/_pickle.c]: We should probably do the same in the get/set methods.

Comment (erlend-aasland) [reply Modules/_pickle.c]: Done.

File: Modules/_pickle.c [hunk 5]
```diff
@@ -7380,7 +7380,7 @@ UnpicklerMemoProxy_New(UnpicklerObject *unpickler)
 {
     UnpicklerMemoProxyObject *self;
 
-    PickleState *state = _Pickle_GetStateByClass(Py_TYPE(unpickler));
+    PickleState *state = _Pickle_FindStateByType(Py_TYPE(self));
```
Comment (kumaraditya303) [inline Modules/_pickle.c]:
```suggestion
    PickleState *state = _Pickle_FindStateByType(Py_TYPE(unpickler));
```

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Modules/_pickle.c (hunk 5 above)
Comment (erlend-aasland) [reply Modules/_pickle.c]: Doh 🤦‍♂️

File: Modules/_pickle.c [hunk 6]
```diff
@@ -5049,47 +5066,26 @@ static PyGetSetDef Pickler_getsets[] = {
...
+    {Py_tp_new, PyType_GenericNew},
+    {Py_tp_alloc, PyType_GenericAlloc},
+    {Py_tp_free, PyObject_GC_Del},
+    {0, NULL},
+};
+
+static PyType_Spec pickler_type_spec = {
+    .name = "_pickle.Pickler",
+    .basicsize = sizeof(PicklerObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_pickle.c]: Missing immutable flag.

File: Modules/_pickle.c [hunk 7]
```diff
@@ -7554,47 +7513,25 @@ static PyGetSetDef Unpickler_getsets[] = {
...
+    {Py_tp_alloc, PyType_GenericAlloc},
+    {Py_tp_new, PyType_GenericNew},
+    {Py_tp_free, PyObject_GC_Del},
+    {0, NULL},
+};
+
+static PyType_Spec unpickler_type_spec = {
+    .name = "_pickle.Unpickler",
+    .basicsize = sizeof(UnpicklerObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_pickle.c]: Same

**Review by kumaraditya303 — APPROVED** [review]

**Comment by kumaraditya303** [comment]
Merged, will start reviewing others in queue.

**Comment by erlend-aasland** [comment]
Thanks for the review, Kumar, and thanks for the initial work, Mohamed!

**Pull request merged**